File: Python_Solns/Trees/maxDepth.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class maxDepth(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """

        #1. base case
        if(root == None):
            return 0
        
        #2. get the right and left subtree heights
        right = self.maxDepth(root.right)
        left = self.maxDepth(root.left)

        #3. return the max height
        return max(right,left)+1

File: Trees/maxDepth.java
package Trees;

import javax.swing.tree.TreeNode;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class maxDepth {
    public int maxDepth(TreeNode root) {

        //1. if the tree is null, then return the null
        if (root == null){
            return 0;
        }

        //2. find the height
        int leftHeight = this.maxDepth(root.left);
        int rightHeight = this.maxDepth(root.right);

        //3. return the max height
        return Math.max(leftHeight, rightHeight)+1;
        
    }
}